### PR TITLE
hcm: add match_upstream to SchemeHeaderTransformation

### LINKED ISSUE #33020: Add variant to SchemeHeaderTransformation to set scheme based on upstream transport
*Description*:
With an additional field in [SchemeHeaderTransformation](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-msg-config-core-v3-schemeheadertransformation) `use_upstream_scheme`, the Router HTTP filter will use the upstream transport socket's protocol to set the scheme.

Examples:
* `client --HTTP--> envoy --HTTPS--> server`:  `:scheme == https`
* `client --HTTPS--> envoy --HTTP--> server`: `:scheme == http`

*Implementation*:
Add an optional boolean to `HttpConnectionManagerConfig` and set scheme to `host->transportSocketFactory().implementsSecureTransport()` if that boolean is defined and is true

*Relevant Links*:
* [PR](https://github.com/envoyproxy/envoy/pull/19583) removed deprecated field `envoy.reloadable_features.preserve_downstream_scheme`
* Mitigation for [this](https://github.com/envoyproxy/envoy/issues/18095) issue no longer applies
* Affects requests to C# grpc servers as demonstrated in [this](https://github.com/istio/istio/issues/34448) issue

### PR DESCRIPTION
**Pull request by wtzhang23**
Commit Message: add match_upstream to SchemeHeaderTransformation
Additional Description: Define a new variant that configures the hcm to override the scheme with the upstream transport protocol. The value is accessible by all downstream filters and is specifically used by the Router filter. Renamed variant from `use_upstream_scheme` to be clear that the scheme matches the transport protocol.
Risk Level: low
Testing: unit testing
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Fixes #33020

### CODE REVIEW

**Comment by wtzhang23** [comment]
Small Qs:
* Not familiar with the async client implementation. Is it possible that especially the gRPC client may take advantage of these settings?
* What setting should the Admin interface use? I have it set to true which from my understanding will be a no-op.

File: api/envoy/config/core/v3/protocol.proto
Comment (adisuissa) [inline api/envoy/config/core/v3/protocol.proto:655]: We are currently discouraging `oneof` fields to allow control planes to be backward-compatible (see https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md).
I suggest adding the field outside of the oneof, and clarify the precedence of the fields.

**Review by adisuissa — COMMENTED** [review]
Thanks!
High-level API comment, but otherwise API LGTM.

File: api/envoy/config/core/v3/protocol.proto
Comment (wtzhang23) [reply api/envoy/config/core/v3/protocol.proto]: Addressed, but unsure what the precedence should be. Made the original field `scheme_to_overwrite` take precedence since it is more explicit.

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [inline api/envoy/config/core/v3/protocol.proto:653]: why not have a transformation which is set_to_match_upstream?

Comment (wtzhang23) [reply api/envoy/config/core/v3/protocol.proto:653]: See resolved conversation above

File: source/server/admin/admin.h
Comment (adisuissa) [inline source/server/admin/admin.h:498]: I'm wondering whether this will break backward-compatibility, and should be false.

File: test/common/http/conn_manager_impl_test_2.cc
Comment (adisuissa) [inline test/common/http/conn_manager_impl_test_2.cc:4268]: Are these counters related to the added feature? If not, can you please remove them

File: test/common/http/conn_manager_impl_test_2.cc
Comment (adisuissa) [inline test/common/http/conn_manager_impl_test_2.cc:4213]: High-level comment:
it seems that the relevant part of the test is essentially:
`EXPECT_TRUE(filter->callbacks_->streamInfo().shouldSchemeMatchUpstream());`

Can this test be refactored to minimize the non-relevant parts?

File: test/common/http/conn_manager_impl_test_base.h
Comment (adisuissa) [inline test/common/http/conn_manager_impl_test_base.h:239]: nit: If this is not being overwritten, consider converting to const.

**Review by adisuissa — COMMENTED** [review]
Thanks!
Please also add a release note.
/assign-from @envoyproxy/senior-maintainers

File: test/common/http/conn_manager_impl_test_base.h
Comment (wtzhang23) [reply test/common/http/conn_manager_impl_test_base.h:239]: Need this to be non-const to override in the `PassMatchUpstreamSchemeHintToStreamInfo` test I wrote.

File: source/common/http/async_client_impl.cc
Comment (wtzhang23) [reply source/common/http/async_client_impl.cc]: Set to false

File: source/extensions/filters/network/http_connection_manager/config.cc
Comment (wtzhang23) [reply source/extensions/filters/network/http_connection_manager/config.cc:654]: Added warning w/ test

File: source/server/admin/admin.h
Comment (wtzhang23) [reply source/server/admin/admin.h]: Set to false.

File: test/common/http/conn_manager_impl_test_2.cc
Comment (wtzhang23) [reply test/common/http/conn_manager_impl_test_2.cc:4213]: Cut as much of the test as I'm familiar with. Will need to dive deeper to see if any other cuts can be made.

**Comment by htuch** [comment]
Reassigning to @alyssawilk for data plane senior maintainer coverage.

File: test/common/http/conn_manager_impl_test_2.cc
Comment (wtzhang23) [reply test/common/http/conn_manager_impl_test_2.cc]: Removed

File: source/common/router/router.cc
Comment (alyssawilk) [inline source/common/router/router.cc:127]: so there's a difference in what we pass in here
I think we need to decide if we want to set scheme based on "is upstream ssl" or "is upstream secure"
e.g. ALTS is secure but not TLS

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [inline api/envoy/config/core/v3/protocol.proto:659]: comment doesn't match code as commented below.  Do we want to set ssl over secure transports or only over TLS?

**Review by alyssawilk — COMMENTED** [review]
I think this version is cleaner, thanks for being willing to move over to it!
/wait on comments

File: source/common/router/router.cc
Comment (wtzhang23) [reply source/common/router/router.cc:127]: Looking at the [golang grpc](https://github.com/grpc/grpc-go/blob/0020ccf9d9fc448fcfa099173df4ad330e96326c/internal/transport/http2_client.go#L304-L306) and [alts](https://github.com/grpc/grpc-go/blob/master/credentials/alts/alts.go#L160-L163) source code, it seems like this should only be set when the upstream is using TLS and not ATLS. The [RFC](https://www.rfc-editor.org/rfc/rfc9110.html#name-https-uri-scheme) also seems to couple it with TLS.

File: api/envoy/config/core/v3/protocol.proto
Comment (wtzhang23) [reply api/envoy/config/core/v3/protocol.proto:662]: Not too familiar with how validation of configs is done but my understanding is it uses `protoc-gen-validate` and I'm not sure if that schema supports cross-field rules?

Comment (wtzhang23) [reply api/envoy/config/core/v3/protocol.proto:662]: Should we just use `oneof` then since this new field doesn't take precedence so if a user wants to use this field they have to not set the other field?

File: source/common/router/router.cc
Comment (wtzhang23) [reply source/common/router/router.cc:127]: Decided to keep it simple and just check if it has an SSL context (TLS) to set the scheme to https, otherwise use http.

File: api/envoy/config/core/v3/protocol.proto
Comment (wtzhang23) [reply api/envoy/config/core/v3/protocol.proto:659]: Decided to set it only over TLS to be consistent with the downstream fallback, which also only sets it if `ssl_connection` is not null.

File: test/common/router/router_2_test.cc
Comment (wtzhang23) [reply test/common/router/router_2_test.cc]: Removed new class and moved code to router.cc where all the tests that use `RouterTest` live.

File: api/envoy/config/core/v3/protocol.proto
Comment (adisuissa) [reply api/envoy/config/core/v3/protocol.proto:662]: > Edit: Actually even then it probably isn't a good idea as there may be a new field which takes precedence in the future.

Yeah, that's the challenge with allowing oneof in this case.
I think that explicitly stating the priorities, and (optionally) emitting a warning when the two fields are set together is a decent compromise.

Comment (wtzhang23) [reply api/envoy/config/core/v3/protocol.proto:662]: I think in this case I should keep the PR as is w.r.t. the API and the emitting the warning. Think I can get a sign-off to pass the CI? Or is it blocked on a review meeting?

**Review by alyssawilk — APPROVED** [review]
apologies - this got tagged as waiting due to CI but fixing CI didn't untag.
LGTM.  @adisuissa any final thoughts? if not no I'll merge tomorrow.

**Comment by alyssawilk** [comment]
ah actually needs shephard LGTM

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Pull request merged**
